api/s2n.h

```diff
@@ -436,6 +436,25 @@ extern int s2n_connection_is_session_resumed(struct s2n_connection *conn);
 S2N_API
 extern int s2n_connection_is_ocsp_stapled(struct s2n_connection *conn);
 
+typedef enum {
+    S2N_HANDSHAKE_SIGNATURE_ANONYMOUS,
+    S2N_HANDSHAKE_SIGNATURE_RSA,
```

Comment:
I don't love that they're all prefixed with HANDSHAKE, but I'm guessing changing the name of the internal ones is a pretty big diff?

Reply:
I was nervous making that change, it seemed easier to not collide with the new enum, happy for better naming suggestions

Reply:
Why not expose the internal enums instead of creating new enums here? I don't see any valuable information getting leaked in exposing just the enums. Having two enums - one public and one internal is harder to maintain.

Reply:
I agree re-using the enums would be nicer, and is a good option for the Hash enum, but the signature enum was a bit more complicated.

My reasoning was the following:
1. Keep the sentinel out of the public interface.
    1. This one is super minor, and I think just adding them to the public interface wouldn't be a big deal.
2. The internal signature enum has a dependency on (tls/s2n_tls_parameters.h)[https://github.com/aws/s2n-tls/blob/879d2b1ca1a030ed4caa69ae1b569c1420d151f4/crypto/s2n_signature.h#L17] which I did not want to introduce into the public API. 

Happy to relent on this if it is acceptable to surface that information.

Reply:
The dependency on tls/s2n_tls_parameters.h was to access these values that are public knowledge. We are not leaking any implementation details here. We could just copy these values over as well, if a dependency is not desired. 

https://github.com/aws/s2n-tls/blob/1be9fbcf21262eb180be79b2860ed89cdc12a1b1/tls/s2n_tls_parameters.h#L130-L136

Reply:
I would rather we didn't own public TLS values defines, especially since they're not prefixed with S2N.

Reply:
I suppose as a middle ground, we could move the enums, but not use these macros to encode the value? Or add the S2N prefix to them? They are only being used for the enum from what I could tell.

Reply:
We could just use the values directly in the enum and expose it. 
 ```c
 /* TLS Signature Algorithms - RFC 5246 7.4.1.4.1 */ 
 /* https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-16 */ 
typedef enum {
    S2N_SIGNATURE_ANONYMOUS = 0,
    S2N_SIGNATURE_RSA = 1,
    S2N_SIGNATURE_ECDSA = 3,

    /* Use Private Range for RSA PSS */
    S2N_SIGNATURE_RSA_PSS_RSAE = 224,
    S2N_SIGNATURE_RSA_PSS_PSS
} s2n_signature_algorithm;
```

Reply:
if exposing internal dependency is not an issue:

- rename and exposing a single set of enums is prob the ideal solution. because having a translation layer is a recipe for error.

however since removing sentinels and exposing the internal enums might be too large a diff, we could do it incrementally. expose outer enum with desired name and eventually switch to exposing internal names.

tests also recommened

Reply:
I'd recommend exposing enum values that matches the TLS IANA values as closely as possible, and uses the private range for values not defined by the IANA for both signature and hash enums. That would leak the least amount of internal s2n details into the public API, and would likely be the most forward stable API we can provide.

Our internal signature enum already matches the public IANA values that we'd want to expose, but our `s2n_hash_algorithm` enum doesn't match the IANA values due to MD5_SHA1 being defined as `7` so that `7` can be used as an array index (even though there's no such IANA value). 

I think we should create two new public enums `s2n_tls_signature_algorithm` and `s2n_tls_hash_algorithm`. The first will exactly match what's used internally, and the 2nd will be a new enum only used for this API. We'd likely still want translation functions for both enums just to decouple the internal and external enums. 

Once these new public enums are added, we can delete some of the `#define`'s for the equivalent values in `s2n_tls_parameters.h` and use the ones in the public `api.h` file instead. 

What do others think?

 ```c
 /* TLS Signature Algorithms - RFC 5246 7.4.1.4.1 */ 
 /* https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-16 */ 
typedef enum {
    S2N_TLS_SIGNATURE_ANONYMOUS = 0,
    S2N_TLS_SIGNATURE_RSA = 1,
    S2N_TLS_SIGNATURE_ECDSA = 3,

    /* Use Private Range for RSA PSS since it's not defined there */
    S2N_TLS_SIGNATURE_RSA_PSS_RSAE = 224,
    S2N_TLS_SIGNATURE_RSA_PSS_PSS
} s2n_tls_signature_algorithm;
```

 ```c
 /* TLS Hash Algorithm - https://tools.ietf.org/html/rfc5246#section-7.4.1.4.1 */ 
 /* https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-18 */ 
typedef enum {
  S2N_TLS_HASH_NONE = 0,
  S2N_TLS_HASH_MD5 = 1,
  S2N_TLS_HASH_SHA1 = 2,
  S2N_TLS_HASH_SHA224 = 3,
  S2N_TLS_HASH_SHA256 = 4,
  S2N_TLS_HASH_SHA384 = 5,
  S2N_TLS_HASH_SHA512 = 6,
  
  /* Use Private Range for MD5_SHA1 */
  S2N_TLS_HASH_MD5_SHA1 = 224
} s2n_tls_hash_algorithm;
```

Reply:
Okay i've just pushed pushed these changes. Please let me know if I misunderstood or got it wrong, my allergies are kicking in so I was being a bit dense :)